File: 프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.js
function solution(ingredient) {
    let count = 0;
    const stack = [];
    
    for(let i=0; i<ingredient.length; i++){
        stack.push(ingredient[i]);
        
        if(stack.length >= 4 && stack.slice(-4).join("") === "1231"){
            stack.splice(-4);
            count++;
        }
    }
    return count;
}


// join("")으로 ingredient 합치고 "1231" 있으면 자르고 count++;
// 없으면 continue 
// count 리턴
// includes()와 replace()를 사용했더니 시간 초과발생 -> stack 이용

File: 프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.py
def solution(ingredient):
    result = 0
    stack = []
    
    # 재료를 하나씩 순회
    for i in ingredient:
        stack.append(i)
        
        # 스택의 길이가 4이상이고 뒤의 4개가 1231 이면
        if(len(stack) >= 4 and stack[-4:] == [1,2,3,1]):
            # 결과 1 증가
            result += 1
            #스택에서 제거
            for _ in range(4):
                stack.pop()
    return result
    
    

